File: demo/csharp/solar_time/solar_time.cs
using System;
using System.Linq;
using demo_helper;
using CosineKitty;

namespace solar_time
{
    class Program
    {
        static int Main(string[] args)
        {
            DemoHelper.ParseArgs("solar_time", args, out Observer observer, out AstroTime time);
            double hourAngle = Astronomy.HourAngle(Body.Sun, time, observer);
            double solarTime = (hourAngle + 12.0) % 24.0;
            int millis = (int)Math.Round(solarTime * 3.6e+6);
            int seconds = millis / 1000;
            millis %= 1000;
            int minutes = seconds / 60;
            seconds %= 60;
            int hours = minutes / 60;
            minutes %= 60;
            hours %= 24;
            Console.WriteLine($"True solar time = {solarTime:F4} hours ({hours:00}:{minutes:00}:{seconds:00}.{millis:000})");
            return 0;
        }
    }
}

File: demo/java/src/main/java/io/github/cosinekitty/astronomy/demo/SolarTime.java
package io.github.cosinekitty.astronomy.demo;
import io.github.cosinekitty.astronomy.*;

public class SolarTime {
    /**
     * Display true solar time for a given geographic location at a given time.
     *
     * @param observer
     * The geographic location for which to calculate true solar time.
     *
     * @param time
     * The date and time for which to calculate true solar time.
     */
    public static int run(Observer observer, Time time) {
        double ha = Astronomy.hourAngle(Body.Sun, time, observer);
        double solarTimeHours = (ha + 12.0) % 24.0;
        int milli = (int) Math.round(solarTimeHours * 3.6e+6);
        int second = milli / 1000;
        milli %= 1000;
        int minute = second / 60;
        second %= 60;
        int hour = minute / 60;
        minute %= 60;
        hour %= 24;

        System.out.printf("True solar time = %7.4f hours (%02d:%02d:%02d.%03d)%n", solarTimeHours, hour, minute, second, milli);
        return 0;
    }
}
